### PR TITLE
partition: fix select statement error on table which partitioned by column with floor func

### LINKED ISSUE #42323: v5.4.3   select statement error on table which partitioned by column with floor func
1. table create sql

```sql
CREATE TABLE `test1` (
  `col1` int(11) NOT NULL DEFAULT '0'
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin /*T! SHARD_ROW_ID_BITS=4 PRE_SPLIT_REGIONS=2 */ 
PARTITION BY RANGE (FLOOR(`col1`))
(PARTITION `p2021` VALUES LESS THAN (202200),
 PARTITION `p2022` VALUES LESS THAN (202300),
 PARTITION `p2023` VALUES LESS THAN (202400),
 PARTITION `p2024` VALUES LESS THAN (202500),
 PARTITION `p2025` VALUES LESS THAN (202600),
 PARTITION `p2026` VALUES LESS THAN (202700));
```

2. case sql
v4.0.11 
```
select * from test1 where col1 = 202303;
```
-> get normal result

v5.4.3
```
select * from test1 where col1 = 202303;
```
-> get error
--> ERROR 1105 (HY000): interface conversion: expression.Expression is *expression.Column, not *expression.ScalarFunction

```
select * from test1 where floor(col1) = 202303;
```
-> get normal result

### PR DESCRIPTION
**Pull request by jiyfhust**
### What problem does this PR solve?

Issue Number: close #42323

Problem Summary:

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: planner/core/integration_partition_test.go [hunk 1]
```diff
@@ -1496,3 +1496,20 @@ func TestPartitionProcessorWithUninitializedTable(t *testing.T) {
...
+	tk.MustExec("create database issue42323")
+	defer tk.MustExec("drop database issue42323")
+
+	tk.MustExec("use issue42323")
+	tk.MustExec("set @@session.tidb_partition_prune_mode = 'dynamic';")
+	tk.MustExec(`CREATE TABLE t(col1 int(11) NOT NULL DEFAULT '0' ) PARTITION BY RANGE (FLOOR(col1))(
+			PARTITION p2021 VALUES LESS THAN (202200),
+			PARTITION p2022 VALUES LESS THAN (202300),
+			PARTITION p2023 VALUES LESS THAN (202400))`)
+	tk.MustExec("insert into t values(202303)")
```
Comment (mjonss) [inline planner/core/integration_partition_test.go:1512]: I think there may be an issue with dynamic prune mode if there are no global statistics, so please add a `analyze table t` before the selects.

**Review by mjonss — APPROVED** [review]
LGTM

**Review by Defined2014 — APPROVED** [review]

**Pull request merged**